WIP

maxorax opened the pull request:
MainActivity - главная и единственная активность в приложении
-- ContactListFragment - фрагмент со списком контактов
---- Короткая карточка контакта
------ Фотография контакта
------ Имя контакта
------ Телефон контакта
При клике пользователя на карточку контакта происходит переход на фрагмент деталей контактов. Необходимо в аргументах передавать ID выбранного контакта.
-- ContactDetailsFragment
---- Фотография контакта
---- Имя контакта
---- Телефон контакта
---- 2-й телефон
---- e-mail контакта
---- 2-й e-mail контакта
---- Описание контакта

- Кнопка "Назад"  работает - т.е. с деталей контакта мы можем вернуться обратно на список контактов
- У активности  Toolbar с colorPrimary цветом
- каждый фрагмент меняет заголовок в Toolbar: ContactListFragment -> "Список контактов", ContactDetailsFragment -> "Детали контакта"
- colorPrimaryDark сделал равным colorPrimary
получившиеся hex цвета: colorPrimary = #2466bd ,colorAccent = #7864c4


<img width="389" alt="Снимок экрана 2020-05-03 в 16 35 26" src="https://user-images.githubusercontent.com/37916430/80914364-22e28680-8d5c-11ea-8828-dc194e35e9e6.png">

<img width="389" alt="Снимок экрана 2020-05-03 в 16 35 23" src="https://user-images.githubusercontent.com/37916430/80914371-26760d80-8d5c-11ea-883e-03b49a9c7246.png">

maxorax committed:
Сделано:
MainActivity - главная и единственная активность в приложении
-- ContactListFragment - фрагмент со списком контактов
---- Короткая карточка контакта
------ Фотография контакта
------ Имя контакта
------ Телефон контакта
При клике пользователя на карточку контакта происходит переход на фрагмент деталей контактов.
-- ContactDetailsFragment
---- Фотография контакта
---- Имя контакта
---- Телефон контакта
---- 2-й телефон
---- e-mail контакта
---- 2-й e-mail контакта
---- Описание контакта

- Кнопка "Назад"  работаtт - т.е. с деталей контакта мы можем вернуться обратно на список контактов
- У активности  Toolbar с colorPrimary цветом
- каждый фрагмент меняет заголовок в Toolbar: ContactListFragment -> "Список контактов", ContactDetailsFragment -> "Детали контакта"
-colorPrimaryDark сделал равным colorPrimary
получившиеся hex цвета: colorPrimary = #2466bd ,colorAccent = #7864c4

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/MainActivity.java:
```diff
@@ -1,14 +1,39 @@
 
 public class MainActivity extends AppCompatActivity {
 
+    Fragment fragment = new ContactListFragment();
```
Хранить ссылку на `Fragment` не стоит в активности. Это может привести к утечке памяти выделенной под фрагмент

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/MainActivity.java:
```diff
@@ -1,14 +1,39 @@
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+        addFragment();
```
Возможна ситуация, при которой фрагмент добавится несколько раз. Например после поворота экрана.
Нужно защититься от этого так:
```java
if (savedInstanceState == null) {
  // можно делать транзакцию, т.к. мы точно запускаемся в первый раз
}
```

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
`FrameLayout` - явно не подходит в качестве контейнера для этого экрана. Посмотри в сторону `ConstraintLayout` либо комбинации из вложенных контейнеров разных типов.

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -0,0 +1,103 @@
+        android:layout_height="26dp"
+        android:layout_marginStart="10dp"
+        android:layout_marginTop="172dp"
+        android:layout_marginEnd="10dp"
```
Верстать таким образом - неверно. Нужно стремиться к резиновой верстке, поэтому лучше оперировать `match_parent` и `wrap_content` для android:layout_width и android:layout_height
Это касается всего проекта

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Здесь то же самое

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,50 @@
+    android:layout_height="match_parent"
+    tools:context=".ContactListFragment">
+
+    <FrameLayout
```
Бессмысленная вложенность контейнеров

maxorax replied:
А что можно использовать для короткой карточки, вместо контейнера?

goblinr replied:
Не заметил, что ты используешь его в качестве фона. Теоретически, можно фон добавить верхнему контейнеру и задать ему отступы

maxorax replied:
У меня к этому контейнеру привязан onClick, если его уберу, то мне к фотке привязать?

goblinr replied:
Зачем? Ты можешь привязать клик к любому View,  включая верхний контейнер

goblinr replied:
Бессмысленная вложенность осталась. Нужно избавиться от этой вложенности

goblinr reviewed (changes requested):
Нужно поправить комментарии и доделать навигацию в приложении
Так же не забудь, что нужно передавать ID выбранного контакта в фрагмент деталей контакта.

maxorax committed:
Сделано:
MainActivity - главная и единственная активность в приложении
-- ContactListFragment - фрагмент со списком контактов
---- Короткая карточка контакта
------ Фотография контакта
------ Имя контакта
------ Телефон контакта
При клике пользователя на карточку контакта происходит переход на фрагмент деталей контактов.
-- ContactDetailsFragment
---- Фотография контакта
---- Имя контакта
---- Телефон контакта
---- 2-й телефон
---- e-mail контакта
---- 2-й e-mail контакта
---- Описание контакта

- Кнопка "Назад"  работаtт - т.е. с деталей контакта мы можем вернуться обратно на список контактов
- У активности  Toolbar с colorPrimary цветом
- каждый фрагмент меняет заголовок в Toolbar: ContactListFragment -> "Список контактов", ContactDetailsFragment -> "Детали контакта"
-colorPrimaryDark сделал равным colorPrimary
получившиеся hex цвета: colorPrimary = #2466bd ,colorAccent = #7864c4

Исправил комментарии

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -1,103 +1,113 @@
+    <ImageView
+        android:id="@+id/foto2"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
```
Для фото лучше оставить жесткие размеры

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -1,103 +1,113 @@
+        android:layout_marginStart="10dp"
+        android:layout_marginTop="20dp"
+        android:layout_marginEnd="10dp"
+        android:paddingLeft="7dp"
```
нечетные размеры.
Это нужно поправить везде

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -8,43 +8,49 @@
-            android:layout_width="67dp"
-            android:layout_height="73dp"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
```
Нужно оставить жесткие размеры для картинки

maxorax committed:
Сделано:
MainActivity - главная и единственная активность в приложении
-- ContactListFragment - фрагмент со списком контактов
---- Короткая карточка контакта
------ Фотография контакта
------ Имя контакта
------ Телефон контакта
При клике пользователя на карточку контакта происходит переход на фрагмент деталей контактов.Необходимо в аргументах передавать ID выбранного контакта
-- ContactDetailsFragment
---- Фотография контакта
---- Имя контакта
---- Телефон контакта
---- 2-й телефон
---- e-mail контакта
---- 2-й e-mail контакта
---- Описание контакта

- Кнопка "Назад"  работаtт - т.е. с деталей контакта мы можем вернуться обратно на список контактов
- У активности  Toolbar с colorPrimary цветом
- каждый фрагмент меняет заголовок в Toolbar: ContactListFragment -> "Список контактов", ContactDetailsFragment -> "Детали контакта"
-colorPrimaryDark сделал равным colorPrimary
получившиеся hex цвета: colorPrimary = #2466bd ,colorAccent = #7864c4

Исправил комментарии

maxorax reviewed (commented):
Поправил комментарии, смог разобраться с bundle и теперь передаю ID через аргументы

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactDetailsFragment.java:
```diff
@@ -0,0 +1,73 @@
+    private TextView phone2;
+    private TextView email;
+    private TextView email2;
+    private TextView description;
```
Все ссылки на View следует обнулять в `onDestroyView()`

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,64 @@
+    private int id;
+    private ImageView contactImage;
+    private TextView name;
+    private TextView phone;
```
Ссылки на  view следует обнулять в onDestroyView()

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public static ContactListFragment newInstance(Contact contact){
```
`ContactListFragment`у не нужно передавать аргументы по условиям ДЗ

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/pojo/Contact.java:
```diff
@@ -0,0 +1,111 @@
+    private String phone2;
+    private String email;
+    private String email2;
+    private String description;
```
`final` ко всем  полям

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/pojo/ContactList.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ContactList {
```
А этот класс зачем? Он нигде не используется.
Нужно удалить.

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -0,0 +1,113 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/email_2" />
+
+
+
+
+
```
Много пустых строк

goblinr commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -0,0 +1,113 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:srcCompat="@mipmap/ic_launcher_round" />
+
+
```
Много пустых строк

goblinr commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,21 @@
+        android:layout_height="match_parent"
+        android:orientation="horizontal">
+    </FrameLayout>
+
+
+</LinearLayout>
```
Много пустых строк

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,58 @@
+        android:layout_marginStart="10dp"
+        android:layout_marginTop="10dp"
+        android:layout_marginEnd="10dp"
+        android:onClick="onClickOnContact"
```
Не делай так.
устанавливай ClickListener в Java коде

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,58 @@
+            android:layout_marginStart="10dp"
+            app:srcCompat="@drawable/ic_launcher_background" />
+    </androidx.cardview.widget.CardView>
+
+
+
+
+
+
+
```
Много пустых строк

goblinr commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,58 @@
+    tools:context=".ContactListFragment">
+
+
+    <androidx.cardview.widget.CardView
```
Не верное решение, должно быть так:
```xml
<androidx.cardview.widget.CardView ...>
    <androidx.constraintlayout.widget.ConstraintLayout ...>
           ...
    </androidx.constraintlayout.widget.ConstraintLayout>
</androidx.cardview.widget.CardView>
```

maxorax committed:
Исправил комментарии

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,43 @@
+    try {
+        ((ContactClick)getActivity()).onClickOnContact(getActivity().findViewById(R.id.contactCard));
+    }
+    catch (ClassCastException e){}
```
Пустой блок  `catch` - так делать нельзя.

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void onClick(View v) {
+    try {
+        ((ContactClick)getActivity()).onClickOnContact(getActivity().findViewById(R.id.contactCard));
```
Лучше сделай как я рассказывал в лекции - проверь оператором `instanceof` в `onAttach` удовлетворяет ли активность к которой присоединяется фрагмент, интерфейсу `ContactClick` и сохрани в поле. Не забудь обнулить в `onDetach`

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,43 @@
+
+
+
+public class ContactListFragment extends Fragment implements View.OnClickListener{
```
Фрагмент реализует интерфейс `View.OnClickListener`, но не устанавливается в качестве слушателя ни одной View

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        onClick(view);
```
Это абсолютно не понятно. Тому кто прочитает данный код - будет казаться, что здесь вызывается метод обратного вызова `onClick`, хотя на самом деле это не так.

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/MainActivity.java:
```diff
@@ -1,14 +1,48 @@
     }
+
+    @Override
+    public void onClickOnContact(View view) {
```
Зачем передавать View в качестве параметра, когда можно сразу передать ID выбранного контакта?

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/pojo/Contact.java:
```diff
@@ -0,0 +1,78 @@
+
+public class Contact {
+
+    private final int ID;
+    private final int IMAGE;
+    private final String NAME;
+    private final String PHONE;
+    private final String PHONE2;
+    private final String EMAIL;
+    private final String EMAIL2;
+    private final String DESCRIPTION;
```
Названия полей - как у констант, хотя по факту - они не константы.

maxorax committed:
Исправил комментарии

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -10,9 +11,17 @@
 
+public class ContactListFragment extends Fragment{
 
+    private Context activity = null;
```
`private ContactClick contactClick = null;`

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -10,9 +11,17 @@
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        if (context instanceof ContactClick){
+            activity = context;
```
`contacktClick = (ContactClick) context;`

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -29,15 +38,27 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
+        activity = null;
+    }
+
+    private void Click() {
```
Название метода не соответствует тому, что в нем происходит.

goblinr replied:
+ название метода с заглавной буквы - не принято в  Java

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/MainActivity.java:
```diff
@@ -34,15 +29,14 @@ private void addFragment(){
-                }
-            });
+    public void onClickOnContactCard(int id) {
+        findViewById(id).setOnClickListener(new View.OnClickListener() {
```
Это полный бред. `setOnClickListener` нужно делать внутри `ContactListFragment` а сюда уже передавать ID  выбранного контакта

maxorax committed:
Исправил комментарии:
setOnClickListener теперь внутри ConatctListFragment
метод Click убрал во все
Исправил
Context activity => ContactClick contactClick
activity = context => contactClick = (ContactClick) context

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,69 @@
+
+public class ContactListFragment extends Fragment{
+
+    private Contact contact;
```
Сделай этот контакт константой

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/ContactListFragment.java:
```diff
@@ -0,0 +1,69 @@
+        contactCard.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+             try {
+             contactClick.onClickOnContactCard(contact);
+            }
+             catch (ClassCastException e){
```
Зачем отлавливать здесь `ClassCastException`?

goblinr replied:
Нужно проверить `contactClick` на `null` прежде чем вызывать на нем метод

goblinr commented on app/src/main/java/android/develop/a65apps/maxim/orlov/MainActivity.java:
```diff
@@ -1,14 +1,35 @@
+
+    @Override
+    public void onClickOnContactCard(Contact contact) {
+                Fragment contactDetails = ContactDetailsFragment.newInstance(contact);
+                FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
+                transaction.replace(R.id.fragments_container, contactDetails).addToBackStack(null).commit();
```
Много табуляций

maxorax committed:
Исправил комментарии:
в ContactListFragment
Contact contact теперь константа,
в onStart заменил try-catch на условие проверки contackClick на null.
В MainActivity убрал лишнюю табуляцую

goblinr reviewed (approved):
Обнови скриншоты
До-оформи PR
И можно мержить
